### PR TITLE
feat: Implement trait dispatch in the comptime interpreter

### PR DESCRIPTION
**Pull request by jfecher**
# Description

## Problem\*

Resolves https://github.com/noir-lang/noir/issues/4925

## Summary\*

Implements trait dispatch in the interpreter - which includes operator overloading.

## Additional Context

I've tried to share as much code as I can between the interpreter and monomorphization.

Most of this PR is rather straightforward adapting code in the monomorphizer to use in the interpreter. The main new bit of code is a stack of `FunctionContext`s in the elaborator. I found at that when having a comptime block in the middle of a function, even a known trait would panic that it had no impl. This was because we previously delayed solving impls to the very end of a function when types were known. I've had to change this into a stack instead so that we can solve impls that were done within a comptime block at the end of that comptime block, just before interpreting so that they'd be defined. Similarly, the `type_variables` list also needed to be placed here since defaulting types can cause some trait constraints to succeed/fail without it. I settled on `function_context` for the name of this stack but am not sure if it fits the case where we have a stack of comptime contexts instead. Similarly, `ComptimeContext` also didn't make sense for the more common case of only having one of them for a function with no comptime blocks.

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_frontend/src/elaborator/mod.rs [hunk 1]
```diff
@@ -417,12 +406,42 @@ impl<'context> Elaborator<'context> {
...
-        self.type_variables.clear();
         self.interner.update_fn(id, hir_func);
         self.current_function = old_function;
         self.current_item = old_item;
     }
 
+    /// Defaults all type variables used in this function context then solves
+    /// all still-unsolved trait constraints in this context.
+    fn check_and_pop_function_context(&mut self) {
+        if let Some(context) = self.function_context.pop() {
```
Comment (asterite) [inline compiler/noirc_frontend/src/elaborator/mod.rs:417]: Should this panic if `function_context` is empty? (it would mean we forgot to push to it)

Comment (jfecher) [reply compiler/noirc_frontend/src/elaborator/mod.rs:417]: I was trying to avoid panics but I suppose we do want to panic in that case - I'll change it

Comment (jfecher) [reply compiler/noirc_frontend/src/elaborator/mod.rs]: Fixed

File: compiler/noirc_frontend/src/elaborator/mod.rs [hunk 2]
```diff
@@ -417,12 +406,42 @@ impl<'context> Elaborator<'context> {
...
         self.trait_bounds.clear();
-        self.type_variables.clear();
         self.interner.update_fn(id, hir_func);
         self.current_function = old_function;
         self.current_item = old_item;
     }
 
+    /// Defaults all type variables used in this function context then solves
+    /// all still-unsolved trait constraints in this context.
+    fn check_and_pop_function_context(&mut self) {
```
Comment (asterite) [inline compiler/noirc_frontend/src/elaborator/mod.rs:416]: Nitpick: should this be `pop_and_check_function_context`?

**Review by asterite — APPROVED** [review]
Looks great!

(but feel free to ask someone else for a review, I don't know the codebase that much yet but the changes here make sense to me :-))

File: compiler/noirc_frontend/src/elaborator/mod.rs (hunk 2 above)
Comment (jfecher) [reply compiler/noirc_frontend/src/elaborator/mod.rs:416]: Maybe? These read very similarly to me. I think I'll leave the name for now

**Pull request merged**